examples/dom/WebReconciler.re

```diff
@@ -0,0 +1,229 @@
+/**
+  * This is an implementation of a reconciler for DOM elements via js_of_ocaml :
+  * http://ocsigen.org/js_of_ocaml/3.1.0/api/Dom_html
+  *
+  * This is just an example but you could use this to create interesting
+  * CLI apps, with a react-like functional API!
+*/
+
+exception InvalidNodePrimitiveMatchInUpdateInstance;
+
+let str = string_of_int;
+
+module Dom = Js_of_ocaml.Dom;
+module Dom_html = Js_of_ocaml.Dom_html;
+module Js = Js_of_ocaml.Js;
+
+/*
+   Step 1: Define the reconciler template
+ */
+module Reconciler = {
+  type hostElement = Js.t(Dom_html.element);
+  type node = Js.t(Dom_html.element);
+
+  let onStale: EventDom.t(unit) = EventDom.create();
+
+  let insertNode = (~parent: node, ~child: node, ~position as _) => {
+    Dom.appendChild(parent, child);
+    parent;
+  };
+
+  let deleteNode = (~parent: node, ~child: node) => {
+    Dom.removeChild(parent, child);
+    parent;
+  };
+
+  let moveNode = (~parent, ~child as _, ~from as _, ~to_ as _) => {
+    parent;
+  };
+
+  let markAsStale = () => EventDom.dispatch(onStale, ());
+
+  let beginChanges = () => ();
+  let commitChanges = () => ();
+};
+
+/*
+    Step 2: Create the actual reconciler using Brisk_reconciler.Make
+ */
+module JsooReact = Brisk_reconciler.Make(Reconciler);
+
+/*
+   Step 3: Define some native components (aka primitives)
+ */
+let document = Dom_html.window##.document;
+
+module View = {
+  let component = JsooReact.nativeComponent("View");
+
+  let make = children => {
+    component((_: JsooReact.Hooks.empty) =>
+      {
+        make: () => {
+          let node = Dom_html.createDiv(document);
+          node;
+        },
+        configureInstance: (~isFirstRender as _, node) => {
+          node;
+        },
+        children,
+      }
+    );
+  };
+
+  let createElement = (~children, ()) => {
+    JsooReact.element(make(JsooReact.listToElement(children)));
+  };
+};
+
+module Image = {
+  let component = JsooReact.nativeComponent("Image");
+
+  let make = (~src, children) => {
+    component((_: JsooReact.Hooks.empty) =>
+      {
+        make: () => {
+          let node = Dom_html.createImg(document);
+          node##.src := Js.string(src);
+          node |> Dom_html.element;
+        },
+        configureInstance: (~isFirstRender as _, n) => {
+          /* TODO: Proper way to downcast? */
+          let node: Js.t(Dom_html.imageElement) = Obj.magic(n);
```

Comment:
That's the big question... I will add an issue about this.

Reply:
Actually, I was thinking the way I resolved this on the lambda-term example is by adding a variant to wrap the native types. This could work in this case too, no? (or if not, I think keeping the magic constraint to the native components feels kind of not too bad).